File: Leetcode/0001-two-sum/solution.py
# ============================================================
# 0001. Two Sum
# https://leetcode.com/problems/two-sum/
# Difficulty: Easy
# Tags: Array, Hash Table
# ============================================================


class Solution:
    def twoSum(self, nums: list[int], target: int) -> list[int]:
        """
        Optimal Solution (One-pass Hash Map)
        Approach: Look up complement in Hash Map
        Time:  O(n) — single pass through the array
        Space: O(n) — Hash Map stores at most n elements
        """
        # Hash Map: value → index
        # For each element, check if its complement (target - num)
        # has been seen before
        seen = {}

        for i, num in enumerate(nums):
            # Calculate complement
            complement = target - num

            # Is complement in the Hash Map?
            if complement in seen:
                # Found! complement's index is seen[complement], current index is i
                return [seen[complement], i]

            # Add current element to Hash Map
            # It will serve as complement for future elements
            seen[num] = i

        # Per constraints, a solution always exists
        # This line is never reached
        return []


# ============================================================
# Test Cases
# ============================================================

if __name__ == "__main__":
    sol = Solution()
    passed = failed = 0

    def test(name: str, got, expected):
        global passed, failed
        if got == expected:
            print(f"✅ PASS: {name}")
            passed += 1
        else:
            print(f"❌ FAIL: {name}")
            print(f"  Got:      {got}")
            print(f"  Expected: {expected}")
            failed += 1

    # Test 1: Basic case — found in first pair
    test("Basic case", sol.twoSum([2, 7, 11, 15], 9), [0, 1])

    # Test 2: Found in the middle
    test("Found in middle", sol.twoSum([3, 2, 4], 6), [1, 2])

    # Test 3: Duplicate values
    test("Duplicate values", sol.twoSum([3, 3], 6), [0, 1])

    # Test 4: Negative numbers
    test("Negative numbers", sol.twoSum([-1, -2, -3, -4, -5], -8), [2, 4])

    # Test 5: Mixed numbers (negative + positive)
    test("Mixed numbers", sol.twoSum([-3, 4, 3, 90], 0), [0, 2])

    # Test 6: Zero values
    test("Zero values", sol.twoSum([0, 4, 3, 0], 0), [0, 3])

    # Test 7: Large values
    test("Large values", sol.twoSum([1000000000, -1000000000], 0), [0, 1])

    # Results
    print(f"\n📊 Results: {passed} passed, {failed} failed, {passed + failed} total")

File: Leetcode/0001-two-sum/solution.java
import java.util.Arrays;
import java.util.HashMap;
import java.util.Objects;

/**
 * 0001. Two Sum
 * https://leetcode.com/problems/two-sum/
 * Difficulty: Easy
 * Tags: Array, Hash Table
 */
class Solution {

    /**
     * Optimal Solution (One-pass Hash Map)
     * Approach: Hash Map da complement ni qidirish
     * Time:  O(n) — massivni faqat 1 marta o'tamiz
     * Space: O(n) — Hash Map ga eng ko'pi bilan n ta element
     */
    public int[] twoSum(int[] nums, int target) {
        // Hash Map: qiymat → indeks
        // Har bir elementni ko'rganda, uning complementi (target - num)
        // oldin ko'rilganmi tekshiramiz
        HashMap<Integer, Integer> seen = new HashMap<>();

        for (int i = 0; i < nums.length; i++) {
            // Complement ni hisoblash
            int complement = target - nums[i];

            // Complement Hash Map da bormi?
            if (seen.containsKey(complement)) {
                // Topildi! complement ning indeksi seen[complement], hozirgi indeks i
                return new int[]{seen.get(complement), i};
            }

            // Hozirgi elementni Hash Map ga qo'shish
            // Keyingi elementlar uchun complement sifatida ishlatiladi
            seen.put(nums[i], i);
        }

        // Constraint bo'yicha har doim javob mavjud
        // Bu yerga hech qachon kelmaydi
        return new int[]{};
    }

    // ============================================================
    // Test Cases
    // ============================================================

    static int passed = 0, failed = 0;

    static void test(String name, int[] got, int[] expected) {
        if (Arrays.equals(got, expected)) {
            System.out.printf("✅ PASS: %s%n", name);
            passed++;
        } else {
            System.out.printf("❌ FAIL: %s%n  Got:      %s%n  Expected: %s%n",
                name, Arrays.toString(got), Arrays.toString(expected));
            failed++;
        }
    }

    public static void main(String[] args) {
        Solution sol = new Solution();

        // Test 1: Basic case — birinchi juftlikda topiladi
        test("Basic case", sol.twoSum(new int[]{2, 7, 11, 15}, 9), new int[]{0, 1});

        // Test 2: O'rtada topiladi
        test("O'rtada topiladi", sol.twoSum(new int[]{3, 2, 4}, 6), new int[]{1, 2});

        // Test 3: Dublikat qiymatlar
        test("Dublikat qiymatlar", sol.twoSum(new int[]{3, 3}, 6), new int[]{0, 1});

        // Test 4: Salbiy sonlar
        test("Salbiy sonlar", sol.twoSum(new int[]{-1, -2, -3, -4, -5}, -8), new int[]{2, 4});

        // Test 5: Aralash sonlar (manfiy + musbat)
        test("Aralash sonlar", sol.twoSum(new int[]{-3, 4, 3, 90}, 0), new int[]{0, 2});

        // Test 6: Nol qiymatlar
        test("Nol qiymatlar", sol.twoSum(new int[]{0, 4, 3, 0}, 0), new int[]{0, 3});

        // Test 7: Katta qiymatlar
        test("Katta qiymatlar", sol.twoSum(new int[]{1000000000, -1000000000}, 0), new int[]{0, 1});

        // Natija
        System.out.printf("%n📊 Natija: %d passed, %d failed, %d total%n",
            passed, failed, passed + failed);
    }
}
